### PR TITLE
Fix size-oblivious guards for mark_dynamic upper bounds

### LINKED ISSUE #125604: Compiling GPT2 with tensorrt backend results in guard failures.
### 🐛 Describe the bug

I'm trying to compile GPT2 model with dynamic shapes and encountered this guard failure at the end. Any suggestions on how to resolve this issue ? 

### Error logs

Please find the error log here : [full_gpt2.log](https://github.com/pytorch/pytorch/files/15224297/full_gpt2.log)

### Minified repro

```py
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import torch_tensorrt

torch_device = "cuda" if torch.cuda.is_available() else "cpu"
tokenizer = AutoTokenizer.from_pretrained("gpt2")

# add the EOS token as PAD token to avoid warnings
model = AutoModelForCausalLM.from_pretrained("gpt2", pad_token_id=tokenizer.eos_token_id).to(torch_device)
model_inputs = tokenizer('I enjoy walking with my cute dog', return_tensors='pt').to(torch_device)
inputs = model_inputs['input_ids']
pyt_outputs = model(inputs)

torch._dynamo.mark_dynamic(model_inputs['input_ids'], 1, min=2, max=1024)
model.forward = torch.compile(model.forward, backend="tensorrt",
                      dynamic=None,
                      options={"truncate_long_and_double": True,
                               "debug": True,
                               "enabled_precisions": {torch.float},
                               "min_block_size": 1,
                               "torch_executed_ops": {"torch.ops.aten.slice.Tensor"}
                               }
                            )

trt_outputs = model(input_ids=inputs)
```

### Versions

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.3.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] onnx==1.15.0
[pip3] onnx-graphsurgeon==0.3.25
[pip3] onnxruntime==1.16.3
[pip3] onnxscript==0.1.0.dev20240205
[pip3] onnxsim==0.4.35
[pip3] optree==0.10.0
[pip3] pytorch-triton==3.0.0+45fff310c8
[pip3] pytorch-triton-rocm==2.2.0
[pip3] torch==2.3.0+cu121
[pip3] torch_tensorrt==2.3.0.dev0+6f945fa4c
[pip3] torchprofile==0.0.4
[pip3] torchsurgeon==0.1.2
[pip3] torchvision==0.18.0+cu121
[pip3] triton==2.3.0

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186451

mark_dynamic(..., max=...) records a StrictMinMaxConstraint for the
selected tensor dimension. That constraint already documents that the
symbol may rely on the same size-oblivious assumptions used for backed
sizes, but the produced backed SIZE symbol was only added to
ShapeEnv.size_like when the global backed_size_oblivious config was
on.

As a result, a guard_size_oblivious check such as size != 1024 was
emitted as a normal guard for a dimension marked dynamic with max=1024.
Guard production then rejected the guard because the user-specified
range includes 1024, causing the ConstraintViolationError reported by
GPT2/TensorRT dynamic-shape compilation.

Fix this by marking symbolic SIZE dimensions with StrictMinMaxConstraint
as size_like when tensor size symbols are produced. This keeps the
behavior tied to explicit constrained-size dimensions instead of enabling
size-oblivious reasoning globally. I considered extending the newer
_check_is_size(max=...) path, but issue 125604 uses mark_dynamic, so the
root cause is the missing size_like classification for StrictMinMaxConstraint.

Fixes #125604
Generated by my agent

Test Plan:
- python test/dynamo/test_misc.py -k test_guard_size_oblivious_mark_dynamic_upper_bound
- python test/dynamo/test_misc.py -k guard_size_oblivious
- python test/test_dynamic_shapes.py -k backed_size_oblivious
- python test/export/test_export.py -k test_export_slice_unbacked_dim1
- lintrunner torch/fx/experimental/symbolic_shapes.py test/dynamo/test_misc.py
- lintrunner -a (failed only on unrelated existing clang-tidy findings outside this change)

Benchmark Results:
- Command: simple torch.compile(backend='eager') loop with mark_dynamic(x, 1, min=2, max=1024), 5 warmup iterations and 20 measured iterations.
- Before: median 0.022022s, mean 0.022265s, min 0.021240s, max 0.024301s.
- After: median 0.021702s, mean 0.021879s, min 0.020765s, max 0.023880s.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: test/dynamo/test_misc.py [hunk 1]
```diff
@@ -13653,6 +13653,18 @@ def f(x):
         self.assertEqual(f(torch.randn(0)).shape, (1,))
         self.assertEqual(f(torch.randn(2)).shape, (2,))
 
+    def test_guard_size_oblivious_mark_dynamic_upper_bound(self):
```
Comment (laithsakka) [inline test/dynamo/test_misc.py:13656]: just delete the failing test guard_size_oblivious shall be deprecated i was waiting for two released i think i shall delete the API now completely .

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -4723,10 +4723,16 @@ def _produce_dyn_sizes_from_int_tuple(
...
-                config.backed_size_oblivious
-                and isinstance(sym, sympy.Symbol)  # could be static
+                isinstance(sym, sympy.Symbol)  # could be static
                 and symbol_is_type(sym, SymT.SIZE)
+                and (
+                    config.backed_size_oblivious
+                    or isinstance(constraint_dims[i], StrictMinMaxConstraint)
+                )
             ):
                 self.size_like.add(sym)
```
Comment (laithsakka) [inline torch/fx/experimental/symbolic_shapes.py:4737]: size_like shall not be relevant nor  guard_size_oblivious both deperecated and shall be deleted soon.

**Review by laithsakka — CHANGES_REQUESTED** [review]

**Comment by laithsakka** [comment]
left a comment on https://github.com/pytorch/pytorch/issues/125604#issuecomment-4655341754 i think 
backed_size_oblivious or export with unbacked probably handles the issue, if not. more investigation is needed. we shall not silently allow specializations.

File: test/dynamo/test_misc.py (hunk 1 above)
Comment (jansel) [reply test/dynamo/test_misc.py:13656]: My agent says thanks, this is correct. The test is asserting behavior for a deprecated API and the range includes the guarded value, so I am closing this PR instead of trying to preserve this behavior.

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (jansel) [reply torch/fx/experimental/symbolic_shapes.py:4737]: My agent says agreed. Adding this symbol to size_like can hide the branch specialization when the marked range still includes the excluded value, so I am closing the PR rather than keeping this change.

**Comment by jansel** [comment]
My agent says I am closing this PR after the requested-changes review. The proposed fix would allow a graph traced under `size != 1024` to cover a `mark_dynamic(..., max=1024)` range that still includes 1024, which is the specialization the guard is supposed to reject. Suggested user workarounds remain to avoid including 1024 in the range, use `maybe_mark_dynamic` when recompiling at 1024 is acceptable, or use the newer unbacked/export path once available.
